tests/shippers/test_logstash.py

```diff
@@ -42,31 +46,72 @@
+            for event in events:
+                _payload.append(ujson.loads(event))
+
+            expected_event = _dummy_event
+            expected_event["_id"] = "_id"
+            assert _payload == [expected_event, expected_event]
+
+            return 200, {}, "okay"
+
+        def event_id_generator(event: dict[str, Any]) -> str:
+            return "_id"
+
+        url = "http://logstash_url"
+        responses.add_callback(responses.PUT, url, callback=request_callback)
+        logstash_shipper = LogstashShipper(logstash_url=url, max_batch_size=2)
+        logstash_shipper.set_event_id_generator(event_id_generator)
+        logstash_shipper.send(_dummy_event)
+        logstash_shipper.send(_dummy_event)
 
-    @mock.patch("shippers.logstash.requests.Session.put", _429_TOO_MANY_REQUESTS)
-    def test_send_too_many_requests(self) -> None:
-        logstash_shipper = LogstashShipper(logstash_url="http://logstash_url")
-        with self.assertRaisesRegex(RuntimeError, "Errors while sending data to Logstash. Return code 429"):
-            logstash_shipper.send(_dummy_event)
+    @responses.activate
+    def test_send_failures(self) -> None:
+        url = "http://logstash_url"
+        with self.subTest("Does not exceed max_retries"):
+            responses.put(url=url, status=429)
+            responses.put(url=url, status=429)
+            responses.put(url=url, status=429)
+            responses.put(url=url, status=200)
+            logstash_shipper = LogstashShipper(logstash_url=url)
+            assert logstash_shipper.send(_dummy_event) == _EVENT_SENT
+        with self.subTest("Exceeds max retries, replay handler set"):
+            for i in range(_MAX_RETRIES):
+                responses.put(url=url, status=429)
+            responses.put(url=url, status=429)
+            logstash_shipper = LogstashShipper(logstash_url=url)
+            logstash_shipper.set_replay_handler(_dummy_replay_handler)
+            assert logstash_shipper.send(_dummy_event) == _EVENT_SENT
+        with self.subTest("Exceeds max retries, replay handler not set"):
+            for i in range(_MAX_RETRIES):
+                responses.put(url=url, status=429)
+            responses.put(url=url, status=429)
+            logstash_shipper = LogstashShipper(logstash_url=url)
+            assert logstash_shipper.send(_dummy_event) == _EVENT_SENT
 
-    @mock.patch("shippers.logstash.requests.Session.put", _200_OK)
+    @responses.activate
     def test_flush(self) -> None:
-        logstash_shipper = LogstashShipper(logstash_url="http://logstash_url", max_batch_size=2)
+        url = "http://logstash_url"
+        responses.put(url=url, status=200)
+        responses.put(url=url, status=200)
+        logstash_shipper = LogstashShipper(logstash_url=url, max_batch_size=2)
         logstash_shipper.send(_dummy_event)
         assert logstash_shipper._events_batch == [_dummy_event]
         logstash_shipper.flush()
```

Comment:
do we miss a last assert on `logstash_shipper._events_batch == []`?

Reply:
Line below (73) :)

Reply:
lol damn github folding code :D